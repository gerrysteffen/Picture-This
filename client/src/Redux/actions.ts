import { UserType } from '../types';

export const setReload = (toggle: Boolean) => ({
  type: 'SET_RELOAD',
  payload: toggle,
});
export const setLoading = (toggle: Boolean) => ({
  type: 'SET_LOADING',
  payload: toggle,
});
export const setAlert = (
  active: Boolean,
  severity: string,
  message: string
) => ({
  type: 'SET_ALERT',
  payload: {
    active: active,
    alertContent: { severity: severity, message: message },
  },
});
export const setAuth = (toggle: Boolean) => ({
  type: 'SET_AUTH',
  payload: toggle,
});
export const setUser = (user: UserType | null) => ({
  type: 'SET_USER',
  payload: user,
});
export const updateUser = (userAttributes: any) => ({
  type: 'UPDATE_USER',
  payload: userAttributes,
});
export const setIsExisting = (toggle: Boolean) => ({
  type: 'SET_EXISTING',
  payload: toggle,
});
