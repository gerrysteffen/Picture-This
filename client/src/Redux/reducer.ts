import { StateType } from '../types';

const initialState: StateType = {
  user: null,
  reloadRequired: true,
  isLoading: true,
  isAuthenticated: false,
  isExistingUser: true,
  activeAlert: false,
  alertContent: {
    severity: 'error',
    message: 'Something went wrong.',
  },
};

const reducer = (
  state: StateType = initialState,
  action: { type: string; payload: any }
) => {
  switch (action.type) {
    case 'SET_AUTH':
      return {
        ...state,
        isAuthenticated: action.payload,
      };
    case 'SET_EXISTING':
      return {
        ...state,
        isExistingUser: action.payload,
      };
    case 'SET_USER':
      return {
        ...state,
        user: action.payload,
      };
    case 'UPDATE_USER':
      return {
        ...state,
        user: {
          ...state.user,
          ...action.payload,
        },
      };
    case 'SET_ALERT':
      return {
        ...state,
        activeAlert: action.payload.active,
        alertContent: action.payload.alertContent
      };
    case 'SET_LOADING':
      return {
        ...state,
        isLoading: action.payload,
      };
    case 'SET_RELOAD':
      return {
        ...state,
        reloadRequired: action.payload,
      };
    default:
      return state;
  }
};

export default reducer;
