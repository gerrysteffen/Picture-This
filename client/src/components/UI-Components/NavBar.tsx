import * as React from 'react';
import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';
import Toolbar from '@mui/material/Toolbar';
import HomeIcon from '@mui/icons-material/Home';
import LogoutIcon from '@mui/icons-material/Logout';
import { useNavigate } from 'react-router-dom';
import APIs from '../../APIServices/index'
import NotificationDropDown from './NotificationDropDown';
import { UserType } from '../../types';

type NavBarType = {
  setIsAuthenticated(user: boolean):void, 
  currentUser: UserType
}

export default function NavBar({setIsAuthenticated, currentUser}: NavBarType) {
  let navigate = useNavigate();
  const logout = async () => {
    setIsAuthenticated(false)
    // localStorage.removeItem('isAuthenticated');
    await APIs.logout();
    navigate('/');
  };
  const goHome = () => navigate('/');

  return (
    <Box sx={{ flexGrow: 1 }}>
      <AppBar
        sx={{ color: 'black', backgroundColor: 'rgba(223, 225, 226, 0.9)' }}
      >
        <Toolbar>

          <IconButton
            onClick={goHome}
            size='large'
            edge='start'
            color='inherit'
            sx={{ mr: 2 }}
          >
            <HomeIcon />
          </IconButton>

          <Box sx={{ flexGrow: 1 , p:1.5}}>
            <img height="35" src='../picture-this1.png' alt='Logo'/>
          </Box>

          {currentUser && (<NotificationDropDown currentUser={currentUser} />) } 

          <IconButton
            onClick={()=>logout()}
            size="large"
            color="inherit">
            <LogoutIcon />
          </IconButton>

        </Toolbar>
      </AppBar>
    </Box>
  );
}
