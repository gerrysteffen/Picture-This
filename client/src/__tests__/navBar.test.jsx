import React from 'react';
import { render, fireEvent, renderWithProviders } from '@testing-library/react';
import NavBar from '../components/UI-Components/NavBar';
import APIs from '../APIServices/index';
import { createStore } from 'redux';
import reducer from '../Redux/reducer';
// import { useNavigate } from 'react-router-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import { useNavigate } from 'react-router-dom';
import { setUser } from '../Redux/actions';

const mockedNavigate = jest.fn()

jest.mock('../APIServices/index', () => ({
  logout: jest.fn(),
}));
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedNavigate
}));

// const navigate = jest.fn();

describe('NavBar', () => {
  const store = createStore(reducer);
  store.dispatch(setUser({pendingInvite:[{owner: 'Gerry', _id: '101', albumName: 'Hello'}]}))
  
  beforeEach(() => {
    jest.clearAllMocks();
  });
  
  it('should render correctly', () => {
    const { container } = render(
    <Provider store={store}>
      <NavBar />
    </Provider>
    )
    expect(container).toMatchSnapshot();
  });

  it('should go home when clicking on the home icon', () => {
    const { getByTestId } = render(
    <Provider store={store}>
      <NavBar />
    </Provider>
    )
    const homeIcon = getByTestId('home-icon');
    fireEvent.click(homeIcon);
    // Assert that the useNavigate hook is being called with '/'
    expect(mockedNavigate).toBeCalledWith('/');
  });

  it('should log out when clicking on the logout icon', async () => {
    const { getByTestId } = render(
    <Provider store={store}>
      <NavBar />
    </Provider>
    )
    const logoutIcon = getByTestId('logout-icon');
    await fireEvent.click(logoutIcon);
    expect(store.getState().isAuthenticated).toBe(false);
    expect(store.getState().user).toBe(null);
    expect(APIs.logout).toBeCalled();
    expect(mockedNavigate).toBeCalledWith('/');
  });
});
