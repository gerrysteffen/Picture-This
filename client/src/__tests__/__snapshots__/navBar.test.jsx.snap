// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-ydbtcs-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="home-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiBox-root css-1tfqhag"
        >
          <img
            alt="Logo"
            height="35"
            src="../picture-this1.png"
          />
        </div>
        <div
          class="css-e53awj-MuiStack-root"
        >
          <div>
            <button
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-xip06h-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiBadge-root MuiBadge-root css-1c32n2y-MuiBadge-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="NotificationsIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 22c1.1 0 2-.9 2-2h-4c0 1.1.89 2 2 2zm6-6v-5c0-3.07-1.64-5.64-4.5-6.32V4c0-.83-.67-1.5-1.5-1.5s-1.5.67-1.5 1.5v.68C7.63 5.36 6 7.92 6 11v5l-2 2v1h16v-1l-2-2z"
                  />
                </svg>
                <span
                  class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightRectangular MuiBadge-overlapRectangular MuiBadge-colorError MuiBadge-badge css-to4x1n-MuiBadge-badge"
                >
                  1
                </span>
              </span>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="logout-icon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4V5z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
  </div>
</div>
`;
